File: fabric/src/main/java/com/lunabbie/mushroommod/LeftoverFoodItem.java
package com.lunabbie.mushroommod;

import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.util.UseAction;
import net.minecraft.world.World;

public class LeftoverFoodItem extends Item {
    private final Item leftover;

    public LeftoverFoodItem(Settings settings, Item leftover) {
        super(settings);
        this.leftover = leftover;
    }

    @Override
    public ItemStack finishUsing(ItemStack stack, World world, LivingEntity user) {
        ItemStack itemStack = super.finishUsing(stack, world, user);
        if (user instanceof PlayerEntity && ((PlayerEntity) user).getAbilities().creativeMode) {
            return itemStack;
        }
        return new ItemStack(leftover);
    }

    @Override
    public UseAction getUseAction(ItemStack stack) {
        if (leftover.equals(Items.BUCKET) || leftover.equals(Items.BOWL)) return UseAction.DRINK;
        else return super.getUseAction(stack);
    }

}

File: fabric/src/main/java/com/lunabbie/mushroommod/items/util/LeftoverFoodItem.kt
package com.lunabbie.mushroommod.items.util

import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.util.UseAction
import net.minecraft.world.World

open class LeftoverFoodItem(settings: Settings, private val leftover: Item) : Item(settings) {
    override fun finishUsing(stack: ItemStack, world: World, user: LivingEntity): ItemStack {
        val itemStack = super.finishUsing(stack, world, user)
        if (user is PlayerEntity && user.abilities.creativeMode) {
            return itemStack
        }
        return ItemStack(leftover)
    }
    override fun getUseAction(stack: ItemStack): UseAction = when (leftover) {
        Items.BUCKET -> UseAction.DRINK
        Items.BOWL -> UseAction.DRINK
        else -> super.getUseAction(stack)
    }
}
